Porous Flow initial

TaraLaForce opened the pull request:
Refs #6740

jwpeterson committed:
Regold output files for VTK output tests.

This file format now contains the element processor IDs by default,
and correctly outputs the solution in parallel.

Refs libMesh/libmesh#882.

jwpeterson committed:
Update libMesh.

This brings in a fix to the VTKIO class that goes along with the
regolded tests in this PR.

Refs #000.

jwpeterson committed:
Only print "basic functionality" check when C++11 is enabled.

dschwen committed:
Make nucleation recoverable (#6729)

Commit:
Add 3D capability to computeSuperellipsoidGradient Ref #6742

WilkAndy committed:
Start of PorousFlow.

This contains YaqiHack

Refs #6740

dschwen committed:
Remove deprecated disp_x, disp_y, disp_z (#6748)

dschwen committed:
Remove accidentally checked in test (#6755)

dschwen committed:
Move homogenized heat conduction into heat conduction module (#6750)

dschwen committed:
Normalize kernel name (#6750)

dschwen committed:
Reorganize heat_conduction module (#6750)

aeslaughter committed:
Fixes uninitialized coupled variables in initQpStatefulProperties

(closes #6760)

WilkAndy committed:
Round2 porous flow

TaraLaForce committed:
added PorousFlowMaterialViscosityConst

WilkAndy committed:
Merge pull request #4 from TaraLaForce/por_flow_initial_6740

added PorousFlowMaterialViscosityConst

WilkAndy committed:
PorFlow to PorousFlow and VarNames to Dictator

jpek15 committed:
implements a constant vector of relative permeabilities, but doesn't yet get the number of phases properly

WilkAndy committed:
PorousFlowMaterialViscosityConst change VarNames to Dictator

WilkAndy committed:
Merge pull request #5 from jpek15/por_flow_initial_6740

implements a constant vector of relative permeabilities, but doesn't …

WilkAndy committed:
Incorporating Tara, Heather and Jonathon changes

cpgr committed:
Commit flux kernel

WilkAndy committed:
PorousFlowDictator now handles num_phases and num_components.  Various fun error messages added.  PorousFlowMaterialRleativePermeabilityConst outputs vector of num_phases size

cpgr committed:
More additions to advective flux kernel

cpgr committed:
More updates to advective flux kernel

WilkAndy committed:
PorousFlowDictator now handles num_phases and num_components.  Various fun error messages added.

WilkAndy committed:
Merge pull request #6 from cpgr/por_flow_initial_6740

Por flow initial 6740

WilkAndy committed:
Uniformising the various contributions.  This compiles and runs without crashing.

dschwen commented on modules/porous_flow/include/kernels/PorFlowComponentMassTimeDerivative.h:
```diff
@@ -0,0 +1,79 @@
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+
```
empty lines

dschwen commented on modules/porous_flow/include/kernels/PorFlowComponentMassTimeDerivative.h:
```diff
@@ -0,0 +1,79 @@
+
+protected:
+  virtual Real computeQpResidual();
+
```
Using empty lines in this manner does not seem particularly useful to me. IMO they should be used sparingly to visually separate blocks that go together. In this particular file I'd suggest adding `///` comments to document the members as well.

dschwen commented on modules/porous_flow/include/kernels/PorFlowComponentMassTimeDerivative.h:
```diff
@@ -0,0 +1,79 @@
+   * @param wrt_num take the derivative of the residual wrt this PorFlow variable
+   */
+  Real computeQpJac(unsigned int pvar);
+
```
empty line

dschwen commented on modules/porous_flow/include/materials/PorFlowMaterial2PhasePS.h:
```diff
@@ -0,0 +1,73 @@
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+
```
empty lines

dschwen commented on modules/porous_flow/include/materials/PorFlowMaterial2PhasePS.h:
```diff
@@ -0,0 +1,73 @@
+  MaterialProperty<std::vector<std::vector<Real> > > & _dsaturation_dvar;
+
+  virtual void initQpStatefulProperties();
+  virtual void computeQpProperties();
```
The de facto standard seems to be to have member methods first and then member variables,

dschwen commented on modules/porous_flow/src/materials/PorFlowMaterialDensityBuilder.C:
```diff
@@ -0,0 +1,79 @@
+    _ddensity_dvar(declareProperty<std::vector<std::vector<Real> > >("dPorFlow_fluid_phase_density_dvar"))
+{
+  _phase_density.resize(_num_phases);
+  _dphase_density_dvar.resize(_num_phases);
```
these should be moved into the initializer lists

dschwen commented on modules/porous_flow/include/materials/PorFlowMaterialMassFractionBuilder.h:
```diff
@@ -0,0 +1,57 @@
+
+protected:
+
+  unsigned int _num_phases;
```
Should be `const` (no empty line after the access qualifiers)

dschwen commented on modules/porous_flow/include/materials/PorFlowMaterialMassFractionBuilder.h:
```diff
@@ -0,0 +1,57 @@
+  unsigned int _num_components;
+
+  /// The variable names UserObject for the Porous-Flow variables
+  const PorFlowVarNames & _porflow_name_UO;
```
Sooo.... there was that discussion about the `PorFlow` prefix... I know that it is quite bikeshed, but what are you going to go with now?

dschwen commented on modules/porous_flow/include/materials/PorFlowMaterialMassFractionBuilder.h:
```diff
@@ -0,0 +1,57 @@
+  virtual void computeQpProperties();
+
+ private:
+  void build_mass_frac(unsigned int qp);
```
Should be named `buildMassFraction()`

dschwen commented on modules/porous_flow/include/userobjects/PorFlowVarNames.h:
```diff
@@ -0,0 +1,140 @@
+   * the simulation are 'energy pwater pgas shape'
+   * then porflow_var_num(2) = 1
+   */
+  unsigned int porflow_var_num(unsigned int moose_var_num) const;
```
all member function names should be camelCased

dschwen commented on modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C:
```diff
@@ -0,0 +1,95 @@
+}
+
+PorFlowComponentMassTimeDerivative::PorFlowComponentMassTimeDerivative(const InputParameters & parameters) :
+  TimeKernel(parameters),
```
initializer indent is 4 spaces

dschwen commented on modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C:
```diff
@@ -0,0 +1,95 @@
+  unsigned int num_phases = _fluid_density[_i].size();
+  mooseAssert(num_phases == _fluid_saturation[_i].size(), "PorFlowComponentMassTimeDerivative: Size of fluid density = " << num_phases << " size of fluid saturation = " << _fluid_saturation[_i].size() << " but both these must be equal to the number of phases in the system");
+
+  Real mass = 0.;
```
`0.0`

dschwen commented on modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C:
```diff
@@ -0,0 +1,95 @@
+PorFlowComponentMassTimeDerivative::computeQpJacobian()
+{
+  /// If the variable is not a PorFlow variable (very unusual), the diag Jacobian terms are 0
+  if (!(_var_is_porflow_var))
```
inner set of parens is unnecessary

dschwen commented on modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C:
```diff
@@ -0,0 +1,95 @@
+  Real dmass = 0.0;
+  for (unsigned ph = 0; ph < num_phases; ++ph)
+  {
+    dmass += _dfluid_density_dvar[_i][ph][pvar]*_fluid_saturation[_i][ph]*_mass_frac[_i][ph][_component_index]*_porosity[_i];
```
operator spacing (also the repeated use of `+=` should probably be replaced by just letting the term flow over multiple lines and `+` operators)

dschwen commented on modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C:
```diff
@@ -0,0 +1,95 @@
+    dmass += _fluid_density[_i][ph]*_fluid_saturation[_i][ph]*_dmass_frac_dvar[_i][ph][_component_index][pvar]*_porosity[_i];
+    dmass += _fluid_density[_i][ph]*_fluid_saturation[_i][ph]*_mass_frac[_i][ph][_component_index]*_dporosity_dvar[_i][pvar];
+  }
+  return _test[_i][_qp]*dmass/_dt;
```
more operator spacing

dschwen commented on modules/porous_flow/src/materials/PorFlowMaterial2PhasePS.C:
```diff
@@ -0,0 +1,123 @@
+#include "PorFlowMaterial2PhasePS.h"
+
+
+
```
rm 4 of these empty lines

dschwen commented on modules/porous_flow/src/materials/PorFlowMaterial2PhasePS.C:
```diff
@@ -0,0 +1,123 @@
+  _saturation_old[_qp].resize(2);
+
+  /*
+   *  YAQI HACK !!
```
LOL

dschwen commented on modules/porous_flow/src/materials/PorFlowMaterial2PhasePS.C:
```diff
@@ -0,0 +1,123 @@
+    _dporepressure_dvar[_qp][phase].assign(_porflow_name_UO.num_v(), 0.0);
+
+  // _porepressure is only dependent on _phase0_porepressure, and its derivative is 1
+  if (!(_porflow_name_UO.not_porflow_var(_phase0_porepressure_varnum)))
```
remove extra parens

dschwen commented on modules/porous_flow/src/materials/PorFlowMaterialMassFractionBuilder.C:
```diff
@@ -0,0 +1,120 @@
+  if (_num_passed_mf_vars != _num_phases*(_num_components - 1))
+    mooseError("PorFlowMaterialMassFractionBuilder: The number of mass_fraction_vars is " << _num_passed_mf_vars << " which must be equal to num_phases (" << _num_phases << ") multiplied by num_components-1 (" << _num_components - 1 << ")");
+
+  _mf_vars.resize(_num_passed_mf_vars);
```
move all these resizes into the initializer list
`    _mf_vars(_num_passed_mf_vars),`

dschwen commented on modules/porous_flow/src/userobjects/PorFlowVarNames.C:
```diff
@@ -0,0 +1,119 @@
+
+void
+PorFlowVarNames::initialize()
+{}
```
```
{
}
```

jwpeterson committed:
Pre-request xyz in Assembly::buildFE().

This is needed at least in Yak, where they use the (x,y,z) locations
of the quadrature points of the FE returned by Assembly::getFE().

Refs libmesh/libMesh#892.

WilkAndy committed:
PorousFlowMaterial1PhaseP_VG added that does 1phase with a van-Genuchten capillary pressure.

Also utils funtions associated with van-Genuchten as we'll use it for other things too.

YaqiHacky removed from Materials as updated moose now correctly initialises Variables

Two Jacobian tests of the fluid-mass kernel added.

Refs ##6740

WilkAndy committed:
mass03 test added that checks the mass-derivative jacobian for multicomponent single phase

WilkAndy committed:
PorousFlowFluidMass postprocessor added, along with tests of it

WilkAndy committed:
fflux01.i added for Heather

hsheldon committed:
First test to check gravity head is established. Fluid bulk modulus >> pp.

WilkAndy committed:
update before attempting upstreaming of flux including the massfrac

jpek15 committed:
Test added for single phase one dimensional system, with a
pressure pulse applied to the left boundary.

hsheldon committed:
- Added random IC
- Bulk mod = 2 (similar to pp)

WilkAndy committed:
Merge pull request #7 from jpek15/por_flow_initial_6740

Test for single phase one dimensional system

hsheldon committed:
Added post processors to indicate error in the solution

hsheldon committed:
Merge branch 'por_flow_initial_6740' of https://github.com/WilkAndy/moose into por_flow_initial_6740

hsheldon committed:
Created 4 versions of grav01 bulk = 2 or 1000, nz = 1 or 10.

WilkAndy committed:
Merge pull request #8 from hsheldon/por_flow_initial_6740

Gravity head tests

TaraLaForce committed:
added two (broken) Corey relative permeability files

WilkAndy committed:
halfway through advectiveflux

WilkAndy committed:
Merge pull request #9 from TaraLaForce/por_flow_initial_6740

added two (broken) Corey relative permeability files

WilkAndy committed:
Merge branch 'por_flow_initial_6740' of github.com:WilkAndy/moose into por_flow_initial_6740

from tara, jonathon and heather

WilkAndy committed:
tara incorporation

WilkAndy committed:
jacobian of fluid flux checked for single-phase single-component

cpgr committed:
Implement dummy variable names for derivative material. Add test kernel and simple test to demonstrate capability

cpgr committed:
Implement dummy variable names for derivative material. Add test kernel and simple test to demonstrate capability

cpgr committed:
fix conflict

WilkAndy committed:
Merge pull request #10 from cpgr/por_flow_initial_6740

Implement dummy variable names for derivative material. Add test kern…

WilkAndy committed:
AdvectiveFlux tested in 2-component scenario.

WilkAndy committed:
Merge branch 'por_flow_initial_6740' of github.com:WilkAndy/moose into por_flow_initial_6740

get Chris DerivativeMaterial example

WilkAndy committed:
pressure_pulse_1d now runs perfectly but i have not yet golded

WilkAndy committed:
jacobian testing with 3 components instead of just 2

WilkAndy committed:
PorousFlowMaterial2PhasePP_VG added.

Jacobian tests of mass-dot and flux kernels in the 2-phase multi-component setting.

Pressure-pulse problem tested in single phase and 2phase multi-component setting, including doco.

Comments added to PorousFlowMaterialMassFractionBuilder regarding the Yaqi Hacky and how the recent change (#6763) does not fully fix the problem.

Refs #6740

WilkAndy committed:
jacobian tests documented.

Refs #6740

WilkAndy committed:
Gravity-head tests added and documented.

Refs #6740

WilkAndy committed:
Buckley-Leverett single-phase test added along with doco.

Refs #6740

WilkAndy committed:
PorousFlowMaterial1PhaseMD_Gaussian aded.

This is an exmple of how we can use nonlinear variables that aren't just porepressure or saturation: it uses log(mass-density).

Also, a derivative contribution to AdvectiveFlux that i forgot about previously
This meant adding properties to 1PhaseP_VG, 2PhasePP_VG and 2PhasePS.

Tests of the Jacobian for the new log(mass-density) material.
Pressure-pulse test for the new log(mass-density) material.

WilkAndy committed:
PorousFlowMaterial1PhaseMD_Gaussian aded.

This is an exmple of how we can use nonlinear variables that aren't just porepressure or saturation: it uses log(mass-density).

Also, a derivative contribution to AdvectiveFlux that i forgot about previously
This meant adding properties to 1PhaseP_VG, 2PhasePP_VG and 2PhasePS.

Tests of the Jacobian for the new log(mass-density) material.

WilkAndy committed:
Bug fix in MD involving residual and jacobian

Also added bl02.i which is buckley-leverett in the MD formulation.
It gives the correct answer, but doesn't converge that well, and
for larger timesteps wanders off to very large P (at some stage
during the nonlinear iterative process) and gives a +inf
in the density, so then the simulations dies.

WilkAndy committed:
Merge branch 'por_flow_initial_6740' of github.com:WilkAndy/moose into por_flow_initial_6740

Conflicts:
	modules/porous_flow/include/materials/PorousFlowMaterial1PhaseMD_Gaussian.h
	modules/porous_flow/src/materials/PorousFlowMaterial1PhaseMD_Gaussian.C
	modules/porous_flow/tests/jacobian/fflux05.i
	modules/porous_flow/tests/jacobian/fflux06.i
	modules/porous_flow/tests/jacobian/mass06.i
	modules/porous_flow/tests/jacobian/mass07.i

cpgr committed:
Create base class for relative permeability materials and update constant and linear relative permeability materials

cpgr committed:
Change PorousFlowRelativePermeabilityConst to PorousFlowRelativePermeabilityConstant

WilkAndy committed:
AdvectiveFlux change constructor on vector<vector> to work with non-clang compilers

cpgr committed:
Add base class for capillary pressure materials, and implement constant and linear versions

cpgr committed:
Create fluid properties base class material, modify DensityConstantBulk, and implement IdealGas and Methane

WilkAndy committed:
pressure_pulse_1d_3comp different petsc options to avoid nan on Tara's computer

TaraLaForce committed:
change to 3 component pressure pulse test to guarantee convergence

WilkAndy committed:
Merge pull request #11 from TaraLaForce/por_flow_initial_6740

change to 3 component pressure pulse test to guarantee convergence

WilkAndy committed:
EffectiveFluidPressure Material added in preparation for mech-coupling

WilkAndy committed:
mass03 test added - it has a volumetric source

cpgr committed:
Add temperature material property to materials

cpgr committed:
Oops - forgot to resize _temperature_qp

WilkAndy committed:
PorousFlowEffectiveStressCoupling Kernel added in preparation for full mechanical coupling (which will need a few more kernels and materials)

cpgr committed:
Make constant viscosity material derive from fluidproperties base class

cpgr committed:
Update Corey relperm to use derivative material interface base class

WilkAndy committed:
PorousFlowMassVolumetricExpansion Kernel added - now all that is needed for mech coupling is varying porosity, yippee.

cpgr committed:
Update joiners to use optional derivative material interface

cpgr committed:
Add use_qps option to tests as an interim

WilkAndy committed:
Merge pull request #12 from cpgr/por_flow_initial_6740

Changes to use derivative material interface properly

cpgr committed:
libmesh update

WilkAndy committed:
bug fixes in mech-fluid coupling kernels

cpgr committed:
Add water fluid properties, and add tests for methane, water and constant bulk modulus fluid fluid properties

cpgr committed:
Put dummy variable names in Dictator, and make all derivatives refer to them

WilkAndy committed:
Merge pull request #13 from cpgr/por_flow_initial_6740

Add water properties and tests of fluid densities and viscosities

WilkAndy committed:
PorousFlow now depends on TensorMechanics.  One poro-elasticity test added.

cpgr committed:
Remove unused variables so no warning in debug mode. Closes #29

WilkAndy committed:
Merge pull request #30 from cpgr/por_flow_initial_6740

Remove unused variables so no warning in debug mode. Closes #29

WilkAndy committed:
start of non-constant porosity

cpgr committed:
Start of changing property names to explicitly reference nodal (like qp for quadpoint properties). Adds base class for fluid state materials. PorousFlowMaterial1PhaseMD_Gaussian updated, along with joiners, fluid properties and relative permeabilites

cpgr committed:
Update PorousFlowMaterial1PhaseP_VG to explicitly reference nodal properties

WilkAndy committed:
PorosityHM and the kernels it uses appear to be debugged.  FINALLY

cpgr committed:
Update PorousFlowMaterial2PhasePP_VG to explicitly refer to nodal properties

cpgr committed:
Update PorousFlowMaterial2PhasePS to explicitly reference nodal quantities

cpgr committed:
Update PorousFlowFluidMass postprocessor to expect saturation_nodal material property

jpek15 committed:
Point source at constant rate for limited time using Dirac Kernel

cpgr committed:
Update kernels to request nodal properties explicitly

cpgr committed:
Remove TestKernel. Closes #33

cpgr committed:
Also remove TestKernl from PorousFLowApp.C. Closes #33

cpgr committed:
Add brine fluid properties

WilkAndy committed:
Merge pull request #31 from jpek15/por_flow_initial_6740

Point source at constant rate for limited time using Dirac Kernel

WilkAndy committed:
Merge pull request #32 from cpgr/por_flow_initial_6740

Explicitly label nodal material properties. Refers to #28

cpgr committed:
Update fluids to explicitly reference nodal quantities

WilkAndy committed:
Merge pull request #34 from cpgr/por_flow_initial_6740

Explicitly reference nodal properties in fluids. Refs #28

WilkAndy committed:
undrained_oedometer poroelasticity test added

WilkAndy committed:
bug-fix for poro_elasticity/tests

TaraLaForce committed:
New App file including Jonathan's TimeLimitedConstantPointSource file

TaraLaForce committed:
changed density0 to density_P0 in material DensityConstBulk and all .i files.  Change was also made to internal parameter in material Brine.C for consistency

WilkAndy commented:
Please close this PR.  It is outdated and i'll submit a new one shortly.

permcody commented:
Closing in favor of #6850
